server/src/main/java/org/elasticsearch/http/HttpPreRequest.java

```diff
@@ -33,6 +33,19 @@ public interface HttpPreRequest {
      */
     String uri();
 
+    /**
+     * The uri without the query string.
+     */
+    default String rawPath() {
+        String uri = uri();
+        final int index = uri.indexOf('?');
```

Comment:
nit: http fragment also a valid char "#" for path end. We probably dont use them.